### DISCUSSION TITLE
Postgres Roles and User impersonation

### DISCUSSION THREAD

**Discussion by saltcod** [supabase/supabase#19595] [General]
During LWX, we rolled out new way to change Postgres roles from the Studio. You may have never thought about this, but Studio connects to your database just like any other Postgres client. It uses a role that functions like your `service role` key, granting it admin privileges to your database.

We've added a new Role dropdown to select a different Postgres role for your queries in Studio. This will let you test your Row Level Security policies and determine which data each role can access.

What do you like most about it? Least?
Does the feature make sense to you? Have we explained what it does? 
Anything you think could be improved? 
Anything you think is missing?

**Comment by nate-oo** [comment]
Awesome work! I love this! Huge improvement to the database management experience in my book.  

I have always carried a level of worry that I might have missed something in my RLS Policies and this helps dramatically with testing and visualizing.  It builds more confidence and trust into the RLS Policies I set up.

Also, it will help with troubleshooting customer issues faster.  "See what they see" faster.

**Related Additional Idea?**
For columns that are of type uuid, and the uuid is in the user table (meaning it is a user uuid), either

**1.** On hover of a uuid, display the name and/or email of the user the uuid belongs to

or even better yet...

**2.** Add some sort of toggle or menu item called something like "Show user name column" to column label or something that allows someone to choose to artificially add a read only column next to the uuid column that displays the name and/or email of the user the uuid belongs to

**Note:** Maybe this doesn't make sense to add and I am being too lazy as I could technically physically add columns to the tables that has this populated or create a view but my guess is many do not do that and many times just have the uuid columns.  I think it would help people better understand the rows they are looking at in the Table Editor when editing / viewing if you can toggle a read only column on and off or on hover view the name and/or email of the user for each uuid column that is a user uuid. 

Just spit balling here...maybe it sparks some ideas :)

**Comment by chekdotdev** [comment]
Hey there! Loving this GUI-side impersonation. Is there a way to achieve this same impersonation from `supabase-js`? We'd love to be able to impersonate from the admin client.

**Reply by GaryAustin1** [reply]
You should be able to do that now from a server by minting a jwt with a jwt library and your user sub,role,etc claims and then pass it in the authorization header.
Edit:  I guess it depends on what all you want to test as it would be a pain to fill in the entire jwt with metadata, etc.

**Reply by blinjrm** [reply]
This is indeed a great functionality in the studio, but impersonating users could also be useful in webapps/APIs.

An example I have in mind is when a user has an API key for a service and we want to leverage RLS in the backend, by anthenticating as the user owning the API key.

At the moment to reproduce this behaviour we have to (AFAIK):
- create an admin client
- get the user email 
- generate an auth magic link
- send a request to that link
- extract the tokens from the response
- create a new, non-admin client
- set a new auth session with the access and refresh tokens

&rarr; That is a lot of back and forth...

Any chance of having the client minting a new token based on the user_id / `sub`, like:
```js
const { data, error } = await supabase.auth.signInWithUserId({
  userId: 'userId',
})
```

**Reply by j4w8n** [reply]
@blinjrm If you can access the user's UUID, then you can mint your own extremely-short-lived-jwt and use that in the auth > global > headers section when creating your non-admin client. That'll be enough for rls - as long as you're doing a check like `auth.uid() = user_id` or whatever.

**Reply by blinjrm** [reply]
Thanks @j4w8n, I'll experiment with this 👍

**Reply by mansueli** [reply]
You can simulate with Supavisor/pgbouncer connections as well:

```sql
psql "postgres://postgres.nacho:[YOUR-PASSWORD]@aws-0-us-east-1.pooler.supabase.com:6543/postgres" <<EOF
BEGIN;
CALL auth.login_as_user('rodrigo@contoso.com');
SELECT * FROM profiles;
COMMIT;
EOF
```

Convenience functions to impersonate are on [this gist](https://gist.github.com/mansueli/ede3563e5dec3e3d4beb88dcaaf66879).

**Comment by Donnerstagnacht** [comment]
I also like the new feature a lot.

I would like to propose two enhancements:

**1. Show a loading indicator or improve performance**
When I tried the new feature today and switched the role from Postgre to anon, the table editor took around 10 seconds to display the rows anon is able to view (in that case no rows). So, the first few times I reloaded the page and thought it was not working. I think it would nice to indicate when the new user role/view is loading/adopted.

**2. Offer the feature operation based: e.g. for SELECT INSERT; UPDATE; DELETE** 
What I like the most about this feature is that the table editor displays which rows the user can "SELECT" - kind of a test without running a test.

It would be nice to have similar views for the other operations, e.g. that the table editor shows all rows which the user could theoretically DELETE or UPDATE or if the user could INSERT something in the table.

**Comment by erchenger** [comment]
It would be really nice to see the access token in the JWT along with the custom claims inside of that when impersonating in the Studio. Maybe I just haven't found the way to do it yet but would be a huge win for me.
